waku/node/peer_manager/peer_manager.nim

```diff
@@ -619,30 +606,59 @@ proc getNumStreams*(pm: PeerManager, protocol: string): (int, int) =
   return (numStreamsIn, numStreamsOut)
 
 proc pruneInRelayConns(pm: PeerManager, amount: int) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
+  let (inRelayPeers, _) = pm.connectedPeers(WakuRelayCodec)
   let connsToPrune = min(amount, inRelayPeers.len)
 
   for p in inRelayPeers[0..<connsToPrune]:
     asyncSpawn(pm.switch.disconnect(p))
 
 proc connectToRelayPeers*(pm: PeerManager) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
-  let maxConnections = pm.switch.connManager.inSema.size
-  let totalRelayPeers = inRelayPeers.len + outRelayPeers.len
-  let inPeersTarget = maxConnections - pm.outRelayPeersTarget
+  var peersToConnect: seq[RemotePeerInfo]
+  var peersToDisconnect: int
+
+  for shard in pm.wakuMetadata.shards.items:
+    let peers = pm.peerStore.getPeersByShard(uint16(pm.wakuMetadata.clusterId), uint16(shard))
+
+    let connectedInPeers = peers.filterIt(
+      pm.peerStore.isConnected(it.peerId) and it.direction == Inbound)
+    
+    let connectedOutPeers = peers.filterIt(
+      pm.peerStore.isConnected(it.peerId) and it.direction == Outbound)
 
-  # TODO: Temporally disabled. Might be causing connection issues
-  #if inRelayPeers.len > pm.inRelayPeersTarget:
-  #  await pm.pruneInRelayConns(inRelayPeers.len - pm.inRelayPeersTarget)
+    let outPeerDiff = pm.outRelayPeersTarget - connectedOutPeers.len
+    let inPeerDiff = pm.inRelayPeersTarget - connectedInPeers.len
 
-  if outRelayPeers.len >= pm.outRelayPeersTarget:
+    if inPeerDiff <= 0:
+      peersToDisconnect += abs(inPeerDiff)
+
+    if outPeerDiff <= 0:
+      continue
+
+    let connectablePeers = peers.filterIt(
+      (not pm.peerStore.isConnected(it.peerId) and pm.canBeConnected(it.peerId)))
+
+    let length = min(outPeerDiff, connectablePeers.len)
+    peersToConnect.insert(connectablePeers[0..<length])
+
+  await pm.pruneInRelayConns(peersToDisconnect)
+  
+  if peersToConnect.len == 0:
     return
 
-  let notConnectedPeers = pm.peerStore.getNotConnectedPeers().mapIt(RemotePeerInfo.init(it.peerId, it.addrs))
-  let outsideBackoffPeers = notConnectedPeers.filterIt(pm.canBeConnected(it.peerId))
-  let numPeersToConnect = min(outsideBackoffPeers.len, MaxParalelDials)
+  # Even with duplicate, after a couple of iteration the target will be reached
+  peersToConnect.deduplicate()
+
+  let chunks = peersToConnect.len div MaxParallelDials
```

Comment:
can we simplify this a bit? understand the goal but maybe we don't need an special case for `mod MaxParallelDials` (the remainder)

Reply:
If I remove it, we can't connect to less than `MaxParallelDials` peers. That would be weird no?

Reply:
did it quickly but perhaps this can prove my point

```
let test = @[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
let paralelDials = 5

for i in countup(0, test.len, paralelDials):
  echo test[i..min((i+paralelDials), test.len-1)]
  
#@[0, 1, 2, 3, 4, 5]
#@[5, 6, 7, 8, 9, 10]
#@[10, 11, 12]
```

this doesn't handle the last iteration as a special case

@SionoiS

Reply:
> did it quickly, may be missing something, but perhaps this can prove my point
> 
> ```
> let test = @[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
> let paralelDials = 5
> 
> for i in countup(0, test.len, paralelDials):
>   echo test[i..min((i+paralelDials), test.len-1)]
>   
> #@[0, 1, 2, 3, 4, 5]
> #@[5, 6, 7, 8, 9, 10]
> #@[10, 11, 12]
> ```
> 
> this doesn't handle the last iteration as a special case
> 
> @SionoiS

I get it now!